src/installer/BeatPackageCompiler/AgentCustomAction.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Deployment.WindowsInstaller;
+
+namespace Elastic.PackageCompiler.Beats
+{
+    public class AgentCustomAction
+    {
+        [CustomAction]
+        public static ActionResult InstallAction(Session session)
+        {
+            try
+            {
+                // If there are no install args, we stop here
+                // // (the MSI just copied the files, there will be no agent-install)
+                if (string.IsNullOrEmpty(session["INSTALLARGS"]))
+                    return ActionResult.Success;
+
+                string install_args = session["INSTALLARGS"];
+                string install_folder = Path.Combine(session["INSTALLDIR"], session["exe_folder"]);
+                System.Diagnostics.Process process = new System.Diagnostics.Process();
+                process.StartInfo.WorkingDirectory = install_folder;
+                process.StartInfo.FileName = "elastic-agent.exe";
+                process.StartInfo.Arguments = "install -f " + install_args;
+                process.StartInfo.CreateNoWindow = true;
+                process.Start();
+                process.WaitForExit();
+
+                session.Log("Agent install return code:" + process.ExitCode);
+
+                if (process.ExitCode == 0)
+                {
+                    // If agent got installed properly, we can go ahead and remove all the files installed by the MSI (best effort)
+                    new DirectoryInfo(install_folder).Delete(true);
+                }
+
+                return process.ExitCode == 0 ? ActionResult.Success : ActionResult.Failure;
+            }
+            catch
+            {
+                return ActionResult.Failure;
+            }
+        }
+
+        [CustomAction]
+        public static ActionResult UpgradeAction(Session session)
+        {
+            session.Log("Detected an agent upgrade via MSI, which is not supported. Aborting.");
+            return ActionResult.Failure;
+        }
+
+        [CustomAction]
+        public static ActionResult UnInstallAction(Session session)
+        {
+            try
+            {
+                // If there are no (un)install args, we stop here
+                if (string.IsNullOrEmpty(session["INSTALLARGS"]))
+                    return ActionResult.Success;
```

Comment:
I don't think a user should have to provide installargs on uninstall, this appears to create lots of weird behavior in practice like uninstalling the MSI via control panel doesn't actually remove agent?

Reply:
you might have missed the `return ActionResult.Success` ?

Reply:
i see what you mean now, ack, need to fix this